fix(file): don't allow downloading File attached via restricted Attach field

asharkinasuit opened the pull request:
Current behavior is that as long as a user is logged in, they can download any file including those marked private. Crucially, this includes files explicitly attached to Attach fields that they otherwise cannot see. They can do this from the attachments list in the sidebar, from the links in the timeline (comments of type "Attachment") and from the File list.

This PR ensures users cannot download files attached via Attach fields that they have no access to (e.g. due to permlevel), since `is_downloadable` is called to check whether the user is allowed to download the file. It does not address the broader issue of permissions on attachments, only specifically Attach fields.

Related issue: https://github.com/frappe/frappe/issues/5894. This issue seems to call for general permissions on downloads, which is a bit trickier, but uses the example of the custom field's permissions that should now be fixed.

(Old version of this PR: #26832; this PR adds much-needed tests to ensure the fix works as intended without breaking existing behavior. The previous version did not pass all these tests.)

asharkinasuit committed:
fix(file): has_permissions should respect attached_to_field's field permissions

This disables all access to files attached via Attach fields that the user has no access to (e.g. due to permlevel)

asharkinasuit committed:
file: add test cases for has_permission

asharkinasuit committed:
fix(file): properly respect field perms

sumitbhanushali commented:
Looks good but relevant attachment and comment also shouldn't be visible to user who don't have access.

asharkinasuit commented:
> Looks good but relevant attachment and comment also shouldn't be visible to user who don't have access.

Ok I will set that up too.

asharkinasuit committed:
fix(file): fomatting new tests

asharkinasuit committed:
fix: timeline can't show inaccessible attachments

The timeline should not show Comments about files the user has no accces to. This  is achieved by generating the Comments based on permissions of the files, which also gives the opportunity to apply up-to-date formatting like graying out deleted files and showing or hiding the padlock depending on is_private.

Drop old mechanism of adding comments for attached files and replace content on Attachment Removed comments to point to the file's name for easier querying. This also fixes the current items' odd formatting "X added/removed attachment <Added/Removed file.txt>" caused by adding formatting on both frontend and backend.

asharkinasuit committed:
fix(file): sidebar can't show inaccessible attachments

asharkinasuit commented:
Timeline and sidebar should now also hide inaccessible attachments.

Ideally, we should also adjust File's `get_permission_query_conditions` to hide inaccessible files from the listview too, but it seems that would require duplicating the logic from `Meta.get_permitted_fieldnames` and functions it uses in a subquery. That might not be desirable both from a performance standpoint and a maintainability standpoint.

While adjusting the timeline, I also noticed that currently, the attachment items have formatting appended on both front end and back end, leading to odd results like
> user123 added attachment Added file123.txt
> user123 removed attachment Removed file123.txt

Furthermore, an Attachment comment would have fixed content, so if a File's  `is_private` property later changes, the lock icon in the timeline might not match the one in the sidebar. Both these issues are resolved here by generating the content for the Attachment (Removed) comments on-the-fly without changing how the frontend renders them.

Finally, I changed the appearance of the timeline's Attachment (Removed) comments. The file name will use the `text-extra-muted` class to emphasize that it is deleted and no longer accessible. This style is used on both the "added" and "removed" comment, and the "added" comment will not have a link if the file was later removed, since that link would be broken anyway if the file was not duplicated.

asharkinasuit commented:
Will anyone be able to review this PR?

asharkinasuit committed:
Merge branch 'develop' into file_permissions

asharkinasuit committed:
fix: quote is now quote_plus

asharkinasuit committed:
fix: use frappe._dict to honor expectations elsewhere

asharkinasuit committed:
fix(test): clean up properly

akhilnarang commented:
Some tests are failing

asharkinasuit committed:
fix(test): commit cleanup

asharkinasuit commented:
The test seemed to fail because the test doctype didn't get dropped as part of the test class teardown (seems there's another rollback() called in the FrappeTestCase base class). It seems only explicitly calling commit will delete the doctypes, but now the semgrep linter complains about explicitly committing. Would this be an acceptable approach?
